kvbc/test/categorization/kv_blockchain_test.cpp

```diff
@@ -229,6 +229,255 @@ TEST_F(categorized_kvbc, delete_block) {
+  //   ASSERT_EQ(std::get<MerkleUpdatesData>(rb.data.category_updates["merkle"]).kv["merkle_key1"], "merkle_value1");
+  // }
+
+  { ASSERT_THROW(block_chain.getRawBlock(0), std::runtime_error); }
+}
+
+TEST_F(categorized_kvbc, link_state_transfer_chain) {
+  KeyValueBlockchain block_chain{db};
+  detail::Blockchain block_chain_imp{db};
+
+  ASSERT_FALSE(block_chain_imp.loadLastReachableBlockId().has_value());
+  ASSERT_EQ(block_chain.getLastReachableBlockId(), 0);
+  // Add block1
+  {
+    Updates updates;
+    MerkleUpdates merkle_updates;
+    merkle_updates.addUpdate("merkle_key1", "merkle_value1");
+    merkle_updates.addDelete("merkle_deleted");
+    updates.add("merkle", std::move(merkle_updates));
+
+    KeyValueUpdates keyval_updates;
+    keyval_updates.addUpdate("kv_key1", "key_val1");
+    keyval_updates.addDelete("kv_deleted");
+    updates.add("kv_hash", std::move(keyval_updates));
+    ASSERT_EQ(block_chain.addBlock(std::move(updates)), (BlockId)1);
+    ASSERT_EQ(block_chain.getGenesisBlockId(), 1);
+  }
+  // Add block2
+  {
+    Updates updates;
+    MerkleUpdates merkle_updates;
+    merkle_updates.addUpdate("merkle_key2", "merkle_value2");
+    merkle_updates.addDelete("merkle_deleted");
+    updates.add("merkle", std::move(merkle_updates));
+
+    KeyValueUpdates keyval_updates;
+    keyval_updates.addUpdate("kv_key2", "key_val2");
+    keyval_updates.addDelete("kv_deleted");
+    updates.add("kv_hash", std::move(keyval_updates));
+    ASSERT_EQ(block_chain.addBlock(std::move(updates)), (BlockId)2);
+    ASSERT_EQ(block_chain.getLastReachableBlockId(), 2);
+    ASSERT_EQ(block_chain_imp.loadLastReachableBlockId().value(), 2);
+  }
+
+  // Add raw block to an invalid id
+  {
+    categorization::RawBlock rb;
+    ASSERT_THROW(block_chain.addRawBlock(rb, 1), std::invalid_argument);
+  }
+
+  // Add raw block with a gap form the last reachable
+  {
+    categorization::RawBlock rb;
+    block_chain.addRawBlock(rb, 5);
+    ASSERT_TRUE(block_chain.getLastStatetransferBlockId().has_value());
+    ASSERT_EQ(block_chain.getLastStatetransferBlockId().value(), 5);
+
+    ASSERT_EQ(block_chain.getLastReachableBlockId(), 2);
+    ASSERT_EQ(block_chain_imp.loadLastReachableBlockId().value(), 2);
+  }
+
+  {
+    categorization::RawBlock rb;
+    block_chain.addRawBlock(rb, 4);
+    ASSERT_TRUE(block_chain.getLastStatetransferBlockId().has_value());
+    ASSERT_EQ(block_chain.getLastStatetransferBlockId().value(), 5);
+
+    ASSERT_EQ(block_chain.getLastReachableBlockId(), 2);
+    ASSERT_EQ(block_chain_imp.loadLastReachableBlockId().value(), 2);
+  }
+
+  {
+    categorization::RawBlock rb;
+    block_chain.addRawBlock(rb, 6);
+    ASSERT_TRUE(block_chain.getLastStatetransferBlockId().has_value());
+    ASSERT_EQ(block_chain.getLastStatetransferBlockId().value(), 6);
+
+    ASSERT_EQ(block_chain.getLastReachableBlockId(), 2);
+    ASSERT_EQ(block_chain_imp.loadLastReachableBlockId().value(), 2);
+  }
+  // Link the chains
```

Comment:
Nice test! Do we need to also verify the contents (via `getBlock()`) of a raw block after it has been linked?

Reply:
yeah, but I'll do it after wiring the categories to actually produce the updates.